docs/notebooks/failure_ego.pct.py

```diff
@@ -107,42 +108,60 @@ def observer(x):
+# We'll model the data on the objective with a regression model, and the data on which points failed with a classification model. The regression model will be a `GaussianProcessRegression` wrapping a GPflow `GPR`, and the classification model a `VariationalGaussianProcess` wrapping a GPflow `VGP` with Bernoulli likelihood.
 
 # %%
 def create_regression_model(data):
     variance = tf.math.reduce_variance(data.observations)
-    kernel = gpflow.kernels.Matern52(variance=variance, lengthscales=0.2 * np.ones(2,))
+    kernel = gpflow.kernels.Matern52(variance=variance, lengthscales=0.2 * np.ones(2, ))
     gpr = gpflow.models.GPR(astuple(data), kernel, noise_variance=1e-5)
     set_trainable(gpr.likelihood, False)
     return gpr
 
 
 def create_classification_model(data):
-    variance = tf.math.reduce_variance(data.observations)
-    kernel = gpflow.kernels.SquaredExponential(variance, lengthscales=0.1 * np.ones(2,))
+    kernel = gpflow.kernels.SquaredExponential(variance=100., lengthscales=0.2 * np.ones(2, ))
     likelihood = gpflow.likelihoods.Bernoulli()
     vgp = gpflow.models.VGP(astuple(data), kernel, likelihood)
+    set_trainable(vgp.kernel.variance, False)
     return vgp
 
 
 regression_model = create_regression_model(initial_data[OBJECTIVE])
 classification_model = create_classification_model(initial_data[FAILURE])
 
+# %% [markdown]
+# ## Create a custom optimize method
+# The new `NatGradTrainedVGP` class has a custom `optimize` method that alternates between Adam steps to optimize the lengthscales and NatGrad steps to optimize the variational parameters:
+
+# %%
+class NatGradTrainedVGP(VariationalGaussianProcess):
+    def optimize(self):
+        set_trainable(self.model.q_mu, False)
+        set_trainable(self.model.q_sqrt, False)
+        variational_params = [(self.model.q_mu, self.model.q_sqrt)]
+        adam_opt = tf.optimizers.Adam(1e-3)
+        natgrad_opt = NaturalGradient(gamma=0.1)
+
+        for step in range(100):
+            natgrad_opt.minimize(self.model.training_loss, var_list=variational_params)
+            adam_opt.minimize(self.model.training_loss, var_list=self.model.trainable_variables)
+
+# %% [markdown]
+# We'll train the GPR model with an L-BFGS-based optimizer, and the GPC model with the custom algorithm above.
+
+# %%
 models = {
     OBJECTIVE: {
         "model": regression_model,
         "optimizer": gpflow.optimizers.Scipy(),
         "optimizer_args": {"options": dict(maxiter=100)},
     },
-    FAILURE: {
-        "model": classification_model,
-        "optimizer": gpflow.optimizers.Scipy(),
-        "optimizer_args": {"options": dict(maxiter=500)},
-    },
+    FAILURE: NatGradTrainedVGP(classification_model)
```

Comment:
mypy really doesn't know how to handle this, so you can do
```
from trieste.models.config import ModelSpec

models: Dict[str, ModelSpec] = {
    ...
```

Reply:
I don't see what you mean, sorry

Reply:
mypy doesn't know how to infer the type of `models`, so it says it's a `Dict[str, object]`, which is too vague for the `optimize` call later, and mypy rejects it. Above, I annotate `models` so that mypy knows what to look for. Type checkers occasionally need help in proving the code is OK

Reply:
you can see this in the tests, where the mypy stage is failing with

Reply:
that removed the test failure